File: backend/app/utils/unit_conversion.py
"""Unit conversion utilities for the costing engine."""

# Conversion factors to base units
# Mass: base unit is grams (g)
# Volume: base unit is milliliters (ml)
# Count: base unit is pieces (pcs)

MASS_CONVERSIONS = {
    "g": 1.0,
    "kg": 1000.0,
    "mg": 0.001,
    "oz": 28.3495,
    "lb": 453.592,
}

VOLUME_CONVERSIONS = {
    "ml": 1.0,
    "l": 1000.0,
    "cl": 10.0,
    "dl": 100.0,
    "tsp": 4.92892,
    "tbsp": 14.7868,
    "cup": 236.588,
    "fl_oz": 29.5735,
}

COUNT_CONVERSIONS = {
    "pcs": 1.0,
    "dozen": 12.0,
}


def get_unit_category(unit: str) -> str | None:
    """Determine the category of a unit (mass, volume, count)."""
    unit_lower = unit.lower()
    if unit_lower in MASS_CONVERSIONS:
        return "mass"
    if unit_lower in VOLUME_CONVERSIONS:
        return "volume"
    if unit_lower in COUNT_CONVERSIONS:
        return "count"
    return None


def convert_to_base_unit(
    quantity: float, from_unit: str, to_base_unit: str
) -> float | None:
    """
    Convert a quantity from one unit to the ingredient's base unit.

    Returns None if units are incompatible or unknown.
    """
    from_unit_lower = from_unit.lower()
    to_base_lower = to_base_unit.lower()

    # Same unit - no conversion needed
    if from_unit_lower == to_base_lower:
        return quantity

    # Check if both units are in the same category
    from_category = get_unit_category(from_unit_lower)
    to_category = get_unit_category(to_base_lower)

    if from_category is None or to_category is None:
        # Unknown unit - return original quantity
        return quantity

    if from_category != to_category:
        # Incompatible units (e.g., mass to volume)
        return None

    # Get conversion maps based on category
    if from_category == "mass":
        conversions = MASS_CONVERSIONS
    elif from_category == "volume":
        conversions = VOLUME_CONVERSIONS
    else:
        conversions = COUNT_CONVERSIONS

    # Convert: from_unit -> standard base -> to_base_unit
    from_factor = conversions.get(from_unit_lower, 1.0)
    to_factor = conversions.get(to_base_lower, 1.0)

    # Convert to standard base, then to target base unit
    standard_quantity = quantity * from_factor
    return standard_quantity / to_factor

File: frontend/src/lib/unitConversion.ts
/**
 * Unit conversion utilities for the frontend.
 * Mirrors backend logic from backend/app/utils/unit_conversion.py
 *
 * Cross-category mass↔volume conversions use the approximation 1g = 1ml
 * (water density). This is a cooking-practical approximation for liquids.
 */

const MASS_CONVERSIONS: Record<string, number> = {
  g: 1,
  kg: 1000,
  mg: 0.001,
  oz: 28.3495,
  lb: 453.592,
};

const VOLUME_CONVERSIONS: Record<string, number> = {
  ml: 1,
  l: 1000,
  tsp: 4.92892,
  tbsp: 14.7868,
  cup: 236.588,
  fl_oz: 29.5735,
};

const COUNT_CONVERSIONS: Record<string, number> = {
  pcs: 1,
  dozen: 12,
};

// Cross-category extras shown in the unit dropdown
// For a mass unit, also show these volume units (and vice versa)
const MASS_CROSS_UNITS = ['ml', 'l'] as const;
const VOLUME_CROSS_UNITS = ['g', 'kg'] as const;

type UnitCategory = 'mass' | 'volume' | 'count';

function getUnitCategory(unit: string): UnitCategory | null {
  const lower = unit.toLowerCase();
  if (lower in MASS_CONVERSIONS) return 'mass';
  if (lower in VOLUME_CONVERSIONS) return 'volume';
  if (lower in COUNT_CONVERSIONS) return 'count';
  return null;
}

function getConversionsForCategory(category: UnitCategory): Record<string, number> {
  switch (category) {
    case 'mass': return MASS_CONVERSIONS;
    case 'volume': return VOLUME_CONVERSIONS;
    case 'count': return COUNT_CONVERSIONS;
  }
}

/**
 * Convert a quantity from one unit to another.
 * Same-category conversions are exact. Mass↔volume uses 1g = 1ml.
 * Returns null only for unrecognised units.
 */
export function convertUnit(quantity: number, fromUnit: string, toUnit: string): number | null {
  const fromLower = fromUnit.toLowerCase();
  const toLower = toUnit.toLowerCase();

  if (fromLower === toLower) return quantity;

  const fromCategory = getUnitCategory(fromLower);
  const toCategory = getUnitCategory(toLower);

  if (fromCategory === null || toCategory === null) return quantity;

  // Same-category conversion
  if (fromCategory === toCategory) {
    const conversions = getConversionsForCategory(fromCategory);
    const fromFactor = conversions[fromLower] ?? 1;
    const toFactor = conversions[toLower] ?? 1;
    return (quantity * fromFactor) / toFactor;
  }

  // Cross-category mass↔volume: treat 1g = 1ml
  if (
    (fromCategory === 'mass' && toCategory === 'volume') ||
    (fromCategory === 'volume' && toCategory === 'mass')
  ) {
    const fromConversions = getConversionsForCategory(fromCategory);
    const toConversions = getConversionsForCategory(toCategory);
    const fromFactor = fromConversions[fromLower] ?? 1;
    const toFactor = toConversions[toLower] ?? 1;
    // Convert to base unit (g or ml), then across (1g=1ml), then to target unit
    return (quantity * fromFactor) / toFactor;
  }

  return null;
}

/**
 * Convert a unit price from one unit to another (inverse of quantity conversion).
 * E.g. $0.01/g → $10.00/kg (price goes up as unit gets larger).
 * Mass↔volume uses 1g = 1ml.
 */
export function convertUnitPrice(price: number, fromUnit: string, toUnit: string): number | null {
  const fromLower = fromUnit.toLowerCase();
  const toLower = toUnit.toLowerCase();

  if (fromLower === toLower) return price;

  const fromCategory = getUnitCategory(fromLower);
  const toCategory = getUnitCategory(toLower);

  if (fromCategory === null || toCategory === null) return price;

  // Same-category: inverse of quantity conversion
  if (fromCategory === toCategory) {
    const conversions = getConversionsForCategory(fromCategory);
    const fromFactor = conversions[fromLower] ?? 1;
    const toFactor = conversions[toLower] ?? 1;
    return price * (toFactor / fromFactor);
  }

  // Cross-category mass↔volume: treat 1g = 1ml (inverse of quantity)
  if (
    (fromCategory === 'mass' && toCategory === 'volume') ||
    (fromCategory === 'volume' && toCategory === 'mass')
  ) {
    const fromConversions = getConversionsForCategory(fromCategory);
    const toConversions = getConversionsForCategory(toCategory);
    const fromFactor = fromConversions[fromLower] ?? 1;
    const toFactor = toConversions[toLower] ?? 1;
    return price * (toFactor / fromFactor);
  }

  return null;
}

/**
 * Get all units compatible with the given unit.
 * For mass units, also includes ml/l. For volume units, also includes g/kg.
 */
export function getCompatibleUnits(unit: string): string[] {
  const category = getUnitCategory(unit.toLowerCase());
  if (!category) return [unit];
  const same = Object.keys(getConversionsForCategory(category));
  if (category === 'mass') return [...same, ...MASS_CROSS_UNITS];
  if (category === 'volume') return [...same, ...VOLUME_CROSS_UNITS];
  return same;
}
